### PR TITLE
borg extract: add --continue flag

### LINKED ISSUE #1356: resume an extract operation
in master (borg 1.1), borg create has quite nice in-file checkpointing abilities, so you can even backup huge files over instable connections.

but if you want to extract such a huge file over a instable connection, it might break down before completion and there is no resume yet.

there are 2 cases:
a) a bunch of files has already been extracted (each file complete, but not all files you want)
b) extraction of a huge file breaks down in the middle

### PR DESCRIPTION
**Pull request by enkore**
Fixes #1356

### CODE REVIEW

File: src/borg/archive.py [hunk 1]
```diff
@@ -509,8 +524,6 @@ def extract_item(self, item, restore_attrs=True, dry_run=False, stdout=False, sp
                         else:
                             fd.write(data)
                 with backup_io():
-                    pos = fd.tell()
-                    fd.truncate(pos)
```
Comment (enkore) [inline src/borg/archive.py]: gnah. mismerge.

File: src/borg/helpers.py [hunk 2]
```diff
@@ -1769,3 +1769,21 @@ def swidth_slice(string, max_width):
...
+
+    Return (sliced_chunks, prefix_length).
+    """
+    def should_drop(chunk):
+        nonlocal current_length
+        dropped = (current_length + chunk.size) <= maximum_length
+        if dropped:
+            current_length += chunk.size
+        return dropped
+    current_length = 0
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: add an empty line after return?

File: src/borg/helpers.py [hunk 3]
```diff
@@ -1769,3 +1769,21 @@ def swidth_slice(string, max_width):
...
+    sliced_chunks = list(dropwhile(should_drop, chunks))
+    assert current_length >= 0
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: hm? in case chunk.size gets negative?

File: src/borg/testsuite/archiver.py [hunk 4]
```diff
@@ -736,6 +736,22 @@ def test_extract_with_pattern(self):
...
+        self.create_regular_file('file2', size=1024 * 100)
+        self.cmd('create', self.repository_location + '::test', 'input')
+
+        with changedir('output'):
+            self.cmd('extract', self.repository_location + '::test')
+            with open('input/file1', 'ab') as file:
+                file.truncate(1234)
+            with open('input/file2', 'ab') as file:
+                file.write(b'extra bytes')
+            print(self.cmd('extract', self.repository_location + '::test', '--continue'))
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver.py]: print?

File: src/borg/testsuite/archiver.py [hunk 5]
```diff
@@ -736,6 +736,22 @@ def test_extract_with_pattern(self):
...
+            assert os.path.getsize('input/file1') == 1024 * 80
+            assert os.path.getsize('input/file2') == 1024 * 100
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver.py]: even better than just checking length would be checking that content is as original file.

File: src/borg/archive.py [hunk 6]
```diff
@@ -428,7 +429,7 @@ def add_file_chunks(chunks):
         return stats
 
     def extract_item(self, item, restore_attrs=True, dry_run=False, stdout=False, sparse=False,
-                     hardlink_masters=None, stripped_components=0, original_path=None, pi=None):
+                     complete_partial=False, stripped_components=0, hardlink_masters=None, original_path=None, pi=None):
```
Comment (ThomasWaldmann) [inline src/borg/archive.py]: you swapped order for some reason?

File: src/borg/archive.py [hunk 7]
```diff
@@ -496,9 +499,21 @@ def extract_item(self, item, restore_attrs=True, dry_run=False, stdout=False, sp
...
+                        existing_length = fd.tell()
+                        # Slice chunks by current length of the existing file.
+                        chunks, prefix_length = slice_chunks(chunks, maximum_length=existing_length)
+                        # We don't bother extracting fractional chunks. Just seek to a chunk boundary.
+                        fd.seek(prefix_length)
+                        fd.truncate()
+                    if pi:
+                        pi.show(increase=prefix_length)
+                ids = [c.id for c in chunks]
                 for _, data in self.pipeline.fetch_many(ids, is_preloaded=True):
```
Comment (ThomasWaldmann) [inline src/borg/archive.py]: not sure if this causes an issue with the preloaded chunks:
iirc it preloads all chunks of a file when processing its item, but then it fetches either none (existing, full file) or only some (existing, partial file), before it goes into fetch-all mode (non-existing files).

Comment (enkore) [inline src/borg/archive.py]: Good catch, didn't consider that. I made up RemoteRepository.discard_preload for that, which looks about right to me, but makes it hang - for some reason.

File: src/borg/archive.py [hunk 8]
```diff
@@ -511,6 +511,8 @@ def extract_item(self, item, restore_attrs=True, dry_run=False, stdout=False, sp
                         # We don't bother extracting fractional chunks. Just seek to a chunk boundary.
                         fd.seek(prefix_length)
                         fd.truncate()
+                    discarded_chunk_ids = [c.id for c in item.chunks[:len(chunks)]]
```
Comment (ThomasWaldmann) [inline src/borg/archive.py]:
```
discarded_count = len(item.chunks) - len(chunks)
discarded_chunks_ids =  [c.id for c in item.chunks[:discarded_count]]
```

File: src/borg/remote.py [hunk 9]
```diff
@@ -506,6 +506,22 @@ def close(self):
...
+
+        for discard_id in discard_ids:
+            msgids = self.cache.get((discard_id,), [])
+            for msgid in list(msgids):
+                if msgid in self.responses:
+                    # Discard anything we may have already received
+                    self.responses.pop(msgid)
+                    msgids.remove(msgid)
+                else:
+                    # The call was sent out, but no response yet; if we got one, we'll ignore it
```
Comment (ThomasWaldmann) [inline src/borg/remote.py]: ... if we get one...

File: src/borg/archive.py (hunk 7 above)
Comment (enkore) [inline src/borg/archive.py]: This is kinda strange. If I set a breakpoint in RR.discard_ids I'll see that neither cache nor responses contain anything, but still removing the IDs from preload_ids makes it hang.

File: src/borg/archive.py (hunk 8 above)
Comment (enkore) [inline src/borg/archive.py]: Thanks and *ouch* that mistake was actually the cause of the deadlocks M(

File: src/borg/archive.py [hunk 10]
```diff
@@ -476,7 +476,7 @@ def extract_item(self, item, restore_attrs=True, dry_run=False, stdout=False, sp
         except UnicodeEncodeError:
             raise self.IncompatibleFilesystemEncodingError(path, sys.getfilesystemencoding()) from None
         except OSError:
-            pass
+            st = False
```
Comment (ThomasWaldmann) [inline src/borg/archive.py]: None?

File: src/borg/archive.py [hunk 11]
```diff
@@ -499,7 +499,13 @@ def extract_item(self, item, restore_attrs=True, dry_run=False, stdout=False, sp
                     return
                 # Extract chunks, since the item which had the chunks was not extracted
             with backup_io():
-                fd = open(path, 'ab')
+                if complete_partial and st:
```
Comment (ThomasWaldmann) [inline src/borg/archive.py]: ... and st is not None:

**Comment by enkore** [comment]
FYI the reason this was moved to b3 is that with larger files it still hangs in call_many, in the same spot as earlier:

> If I set a breakpoint in RR.discard_ids I see that neither cache nor responses contain anything, but still - removing the IDs from preload_ids makes it hang in call_many, in the while not self.to_send and (calls or self.preload_ids) and len(waiting_for) < MAX_INFLIGHT: loop.

I haven't looked at it again yet.

File: src/borg/archive.py [hunk 12]
```diff
@@ -496,9 +499,30 @@ def extract_item(self, item, restore_attrs=True, dry_run=False, stdout=False, sp
...
+                    with backup_io():
+                        fd.seek(0, io.SEEK_END)
+                        existing_length = fd.tell()
+                        # Slice chunks by current length of the existing file.
+                        chunks, prefix_length = slice_chunks(chunks, maximum_length=existing_length)
+                        # We don't bother extracting fractional chunks. Just seek to a chunk boundary.
+                        fd.seek(prefix_length)
+                        fd.truncate()
+                    discarded_count = len(item.chunks) - len(chunks)
+                    discarded_chunks_ids = [c.id for c in item.chunks[:discarded_count]]
```
Comment (ThomasWaldmann) [inline src/borg/archive.py]: nitpick: discard_count / discard_chunk_ids (not past tense)

**Comment by enkore** [comment]
"later" and no activity = close.

**Pull request closed without merging**